File: typescript/src/backends/otlp.ts
/**
 * OTLP (OpenTelemetry Protocol) backend for telemetry data.
 *
 * Sends telemetry data to an OTLP-compatible endpoint using HTTP/JSON.
 * Uses only Node.js standard library - no external dependencies.
 */

import * as zlib from "zlib";
import { TelemetryBackend } from "./base";

/**
 * Configuration options for OTLP backend.
 * Exported as part of the public API for users who want to use OTLPBackend directly.
 */
export interface OTLPBackendConfig {
  /** OTLP HTTP endpoint (e.g. https://telemetry.namastex.ai/v1/traces) */
  endpoint: string;
  /** HTTP timeout in milliseconds (default: 5000) */
  timeout?: number;
  /** Maximum retry attempts (default: 3) */
  maxRetries?: number;
  /** Retry backoff base delay in milliseconds (default: 1000) */
  retryBackoffBase?: number;
  /** Enable gzip compression (default: true) */
  compressionEnabled?: boolean;
  /** Compression threshold in bytes (default: 1024) */
  compressionThreshold?: number;
  /** Enable verbose console logging for debugging (default: false) */
  verbose?: boolean;
}

/**
 * OTLP HTTP/JSON backend for sending telemetry data.
 *
 * Sends traces, metrics, and logs to an OTLP-compatible endpoint
 * with automatic retry logic and compression support.
 */
export class OTLPBackend implements TelemetryBackend {
  private endpoint: string;
  private timeout: number;
  private maxRetries: number;
  private retryBackoffBase: number;
  private compressionEnabled: boolean;
  private compressionThreshold: number;
  private verbose: boolean;

  constructor(config: OTLPBackendConfig) {
    this.endpoint = config.endpoint;
    this.timeout = config.timeout || 5000;
    this.maxRetries = config.maxRetries || 3;
    this.retryBackoffBase = config.retryBackoffBase || 1000;
    this.compressionEnabled = config.compressionEnabled !== false;
    this.compressionThreshold = config.compressionThreshold || 1024;
    this.verbose = config.verbose || false;
  }

  /**
   * Compress payload using gzip.
   *
   * @param data - Data to compress
   * @returns Compressed buffer
   */
  private async compressPayload(data: string): Promise<Buffer> {
    return new Promise((resolve, reject) => {
      zlib.gzip(Buffer.from(data, "utf-8"), (error, result) => {
        if (error) {
          reject(error);
        } else {
          resolve(result);
        }
      });
    });
  }

  /**
   * Send HTTP request with retry logic.
   *
   * @param endpoint - Target endpoint
   * @param payload - Request payload
   * @param attempt - Current attempt number (internal)
   * @returns True if request succeeded, false otherwise
   */
  private async sendWithRetry(
    endpoint: string,
    payload: unknown,
    attempt: number = 0,
  ): Promise<boolean> {
    try {
      const payloadString = JSON.stringify(payload);
      let body: Buffer | string = payloadString;
      const headers: Record<string, string> = {
        "Content-Type": "application/json",
      };

      // Apply compression if enabled and payload exceeds threshold
      if (
        this.compressionEnabled &&
        Buffer.byteLength(payloadString, "utf-8") > this.compressionThreshold
      ) {
        body = await this.compressPayload(payloadString);
        headers["Content-Encoding"] = "gzip";
      }

      // Send HTTP request using native fetch (Node.js 18+)
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.timeout);
      timeoutId.unref(); // Allow process to exit

      try {
        const response = await fetch(endpoint, {
          method: "POST",
          headers,
          body,
          signal: controller.signal,
        });

        clearTimeout(timeoutId);

        // Check for retryable errors (5xx)
        if (response.status >= 500 && response.status < 600) {
          throw new Error(`Server error: ${response.status}`);
        }

        // Don't retry on 4xx errors (client errors)
        if (response.status >= 400 && response.status < 500) {
          if (this.verbose) {
            console.debug(
              `Telemetry event failed with status ${response.status}`,
            );
          }
          return false; // Don't retry
        }

        if (response.status !== 200) {
          if (this.verbose) {
            console.debug(
              `Telemetry event failed with status ${response.status}`,
            );
          }
          return false;
        }

        return true;
      } catch (error) {
        clearTimeout(timeoutId);
        throw error;
      }
    } catch (error) {
      // Retry logic for network errors or 5xx responses
      if (attempt < this.maxRetries) {
        const backoffDelay = this.retryBackoffBase * Math.pow(2, attempt);
        if (this.verbose) {
          console.debug(
            `Telemetry request failed (attempt ${attempt + 1}/${this.maxRetries + 1}), retrying in ${backoffDelay}ms...`,
          );
        }
        await new Promise((resolve) => {
          const timer = setTimeout(resolve, backoffDelay);
          timer.unref(); // Allow process to exit
        });
        return this.sendWithRetry(endpoint, payload, attempt + 1);
      }

      // Max retries exceeded
      if (this.verbose) {
        console.debug(
          `Telemetry event error after ${this.maxRetries + 1} attempts:`,
          error,
        );
      }
      return false;
    }
  }

  /**
   * Send a trace payload to the OTLP endpoint.
   *
   * @param payload - OTLP trace payload
   * @returns Promise that resolves to true if successful, false otherwise
   */
  async sendTrace(payload: unknown): Promise<boolean> {
    return this.sendWithRetry(this.endpoint, payload);
  }

  /**
   * Send a metric payload to the OTLP endpoint.
   *
   * @param payload - OTLP metric payload
   * @param metricsEndpoint - Optional metrics-specific endpoint
   * @returns Promise that resolves to true if successful, false otherwise
   */
  async sendMetric(
    payload: unknown,
    metricsEndpoint?: string,
  ): Promise<boolean> {
    const endpoint = metricsEndpoint || this.endpoint;
    return this.sendWithRetry(endpoint, payload);
  }

  /**
   * Send a log payload to the OTLP endpoint.
   *
   * @param payload - OTLP log payload
   * @param logsEndpoint - Optional logs-specific endpoint
   * @returns Promise that resolves to true if successful, false otherwise
   */
  async sendLog(payload: unknown, logsEndpoint?: string): Promise<boolean> {
    const endpoint = logsEndpoint || this.endpoint;
    return this.sendWithRetry(endpoint, payload);
  }

  /**
   * Flush any buffered data.
   *
   * OTLP backend does not buffer data - all sends are immediate.
   * This method is provided for interface compatibility and always returns true.
   *
   * @returns Promise that resolves to true
   */
  async flush(): Promise<boolean> {
    // OTLP backend sends data immediately, no buffering
    return true;
  }
}

File: python/src/automagik_telemetry/backends/otlp.py
"""
OTLP backend for telemetry data.

Sends traces, metrics, and logs using the OpenTelemetry Protocol (OTLP) over HTTP.
Uses only standard library - no external dependencies.
"""

import gzip
import json
import logging
import time
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

from .base import TelemetryBackend

logger = logging.getLogger(__name__)


class OTLPBackend(TelemetryBackend):
    """
    OTLP HTTP backend for sending telemetry data.

    Sends OTLP-formatted traces, metrics, and logs to OpenTelemetry collectors
    using HTTP/JSON with optional gzip compression and retry logic.
    """

    def __init__(
        self,
        endpoint: str,
        metrics_endpoint: str,
        logs_endpoint: str,
        timeout: int = 5,
        max_retries: int = 3,
        retry_backoff_base: float = 1.0,
        compression_enabled: bool = True,
        compression_threshold: int = 1024,
        verbose: bool = False,
    ):
        """
        Initialize OTLP backend.

        Args:
            endpoint: OTLP traces endpoint (e.g., https://telemetry.namastex.ai/v1/traces)
            metrics_endpoint: OTLP metrics endpoint (e.g., https://telemetry.namastex.ai/v1/metrics)
            logs_endpoint: OTLP logs endpoint (e.g., https://telemetry.namastex.ai/v1/logs)
            timeout: HTTP timeout in seconds (default: 5)
            max_retries: Maximum number of retry attempts (default: 3)
            retry_backoff_base: Base backoff time in seconds (default: 1.0)
            compression_enabled: Enable gzip compression (default: True)
            compression_threshold: Minimum payload size for compression in bytes (default: 1024)
            verbose: Enable verbose logging (default: False)
        """
        self.endpoint = endpoint
        self.metrics_endpoint = metrics_endpoint
        self.logs_endpoint = logs_endpoint
        self.timeout = timeout
        self.max_retries = max_retries
        self.retry_backoff_base = retry_backoff_base
        self.compression_enabled = compression_enabled
        self.compression_threshold = compression_threshold
        self.verbose = verbose

    def _compress_payload(self, payload: bytes) -> bytes:
        """
        Compress payload using gzip if it exceeds threshold.

        Args:
            payload: Raw bytes to compress

        Returns:
            Compressed or original payload
        """
        if self.compression_enabled and len(payload) >= self.compression_threshold:
            return gzip.compress(payload)
        return payload

    def _send_with_retry(
        self, endpoint: str, payload: dict[str, Any], signal_type: str = "trace"
    ) -> bool:
        """
        Send payload with retry logic and exponential backoff.

        Args:
            endpoint: Target endpoint URL
            payload: OTLP payload to send
            signal_type: Type of signal (trace, metric, log) for logging

        Returns:
            True if successful, False otherwise
        """
        try:
            # Serialize payload
            payload_bytes = json.dumps(payload).encode("utf-8")

            # Compress if needed
            original_size = len(payload_bytes)
            payload_bytes = self._compress_payload(payload_bytes)
            compressed = len(payload_bytes) < original_size

            # Prepare headers
            headers = {"Content-Type": "application/json"}
            if compressed:
                headers["Content-Encoding"] = "gzip"

            # Verbose mode logging
            if self.verbose:
                print(f"\n[Telemetry] Sending {signal_type}")
                print(f"  Endpoint: {endpoint}")
                print(f"  Size: {len(payload_bytes)} bytes (compressed: {compressed})")
                print(f"  Payload preview: {json.dumps(payload, indent=2)[:200]}...\n")

            # Retry loop with exponential backoff
            last_exception = None
            for attempt in range(self.max_retries + 1):
                try:
                    request = Request(endpoint, data=payload_bytes, headers=headers)

                    with urlopen(request, timeout=self.timeout) as response:
                        if response.status == 200:
                            logger.debug(f"OTLP {signal_type} sent successfully")
                            return True
                        elif response.status >= 500:
                            # Server error - retry
                            last_exception = Exception(f"Server error: {response.status}")
                        else:
                            # Client error - don't retry
                            logger.debug(
                                f"Telemetry {signal_type} failed with status {response.status}"
                            )
                            return False

                except (URLError, HTTPError, TimeoutError) as e:
                    last_exception = e
                    # Check if we should retry
                    if isinstance(e, HTTPError) and e.code < 500:
                        # Client error - don't retry
                        logger.debug(f"Telemetry {signal_type} failed: {e}")
                        return False

                # Wait before retry (exponential backoff)
                if attempt < self.max_retries:
                    backoff_time = self.retry_backoff_base * (2**attempt)
                    time.sleep(backoff_time)

            # All retries exhausted
            if last_exception:
                logger.debug(
                    f"Telemetry {signal_type} failed after {self.max_retries} retries: {last_exception}"
                )
            return False

        except Exception as e:
            # Log any other errors in debug mode
            logger.debug(f"Telemetry {signal_type} error: {e}")
            return False

    def send_trace(self, otlp_payload: dict[str, Any]) -> bool:
        """
        Send an OTLP trace payload to the traces endpoint.

        Args:
            otlp_payload: OTLP-formatted trace payload with resourceSpans structure

        Returns:
            True if successful, False otherwise
        """
        return self._send_with_retry(self.endpoint, otlp_payload, "trace")

    def send_metric(self, otlp_payload: dict[str, Any]) -> bool:
        """
        Send an OTLP metric payload to the metrics endpoint.

        Args:
            otlp_payload: OTLP-formatted metric payload with resourceMetrics structure

        Returns:
            True if successful, False otherwise
        """
        return self._send_with_retry(self.metrics_endpoint, otlp_payload, "metric")

    def send_log(self, otlp_payload: dict[str, Any]) -> bool:
        """
        Send an OTLP log payload to the logs endpoint.

        Args:
            otlp_payload: OTLP-formatted log payload with resourceLogs structure

        Returns:
            True if successful, False otherwise
        """
        return self._send_with_retry(self.logs_endpoint, otlp_payload, "log")

    def flush(self) -> bool:
        """
        Flush any pending data.

        Note: OTLP backend sends data immediately, so this is a no-op.
        Included for consistency with other backends.

        Returns:
            True (always succeeds since OTLP doesn't buffer)
        """
        return True
